graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,376 @@
+        PUBLIC,
+        PRIVATE
+    }
+
+    private static class CacheControlState implements InstrumentationState {
+        private final CacheControlPolicy overallPolicy = new CacheControlPolicy();
+
+        public Optional<String> getOverallPolicyString() {
+            return overallPolicy.getPolicy();
+        }
+
+        public void restrictCachePolicy(CacheControlPolicy policy) {
+            this.overallPolicy.restrict(policy);
+        }
+    }
+
+    private static class CacheControlPolicy implements InstrumentationState {
+        private Optional<Integer> maxAge = Optional.empty();
+        private Optional<CacheControlScope> scope = Optional.of(CacheControlScope.PUBLIC);
+
+        void restrict(CacheControlPolicy policy) {
+            if (policy.maxAge.isPresent() && (!maxAge.isPresent() || policy.maxAge.get() < maxAge.get())) {
+                this.maxAge = policy.maxAge;
+            }
+
+            if (policy.scope.isPresent() && (!scope.isPresent() || !scope.get().equals(CacheControlScope.PRIVATE))) {
+                this.scope = policy.scope;
+            }
+        }
+
+        void restrict(Hint hint) {
+            if (hint.maxAge.isPresent() && (!maxAge.isPresent() || hint.maxAge.get() < maxAge.get())) {
+                this.maxAge = hint.maxAge;
+            }
+
+            if (hint.scope.isPresent() && (!scope.isPresent() || !scope.get().equals(CacheControlScope.PRIVATE))) {
+                this.scope = hint.scope;
+            }
+        }
+
+        void restrict(Integer maxAge) {
+            this.maxAge = Optional.of(maxAge);
+        }
+
+        void replace(Hint hint) {
+            if (hint.maxAge.isPresent()) {
+                this.maxAge = hint.maxAge;
+            }
+
+            if (hint.scope.isPresent()) {
+                this.scope = hint.scope;
+            }
+        }
+
+        void replace(Optional<CacheControlScope> scope) {
+            if (scope.isPresent()) {
+                this.scope = scope;
+            }
+        }
+
+        public Optional<String> getPolicy() {
+            Integer maxAgeValue = maxAge.orElse(0);
+            if (maxAgeValue.equals(0)) {
+                return Optional.empty();
+            }
+            return Optional.of(String.format("max-age=%d, %s", maxAgeValue, scope.orElse(CacheControlScope.PUBLIC).toString().toLowerCase()));
+        }
+
+        public boolean hasMaxAge() {
+            return maxAge.isPresent();
+        }
+
+        public boolean hasScope() {
+            return scope.isPresent();
+        }
+    }
+
+    private static class HintCache {
+        private final Map<String, Hint> typeHintCache = new HashMap<>();
+        private final Map<String, Hint> fieldHintCache = new HashMap<>();
+
+        public Optional<Hint> getHint(GraphQLFieldDefinition fieldDefinition, GraphQLCompositeType parentType) {
+            String cacheKey = String.format("%s.%s", parentType.getName(), fieldDefinition.getName());
```

Comment:
Nitpicking: I'm curious if the `fieldDefinition` could be used as cacheKey to avoid passing the `parentType` argument. It feels like they should be unique once the schema is parsed but it's certainly safer to use Strings so no need to change anything.

Reply:
good call! i was in a typescript mindset that favors string keys in maps.